File: JavaScript/4-usage.js
'use strict';

const emitter = require('./3-enhanced.js');

const application = emitter();

application.on('smth', (data) => {
  console.log('Certain event');
  console.dir(data);
});

application.on('*', (name, data) => {
  console.log('Any event');
  console.dir([name, data]);
});

application.emit('smth', { a: 5 });
application.emit('smth2', { a: 500 });
application.emit('*', { a: 700 });

File: Haskell/4-usage.hs
module Main where

import EnhancedEmitter

data Struct = Struct { a :: Int } deriving (Show)

application  = emitter
application' = onAny printAny . on "smth" printCertain $ application

main = mconcat . concat . map ($ application') $
        [ emit "smth" (Struct 5)
        , emit "smth2" (Struct 500)
        , emit "*" (Struct 700)
        ]

printCertain :: Show a => a -> IO ()
printCertain dat = do
  putStrLn "Certain event"
  print dat

printAny :: Show a => String -> a -> IO ()
printAny name dat = do
  putStrLn "Any event"
  print (name, dat)
